pkg/provider/cloud/openstack/tags.go

```diff
@@ -0,0 +1,137 @@
+				if id != "" {
+					idSet.Insert(id)
+				}
+			}
+			idSet.Insert(clusterName)
+			tags[i] = TagPrefixOwnedByCluster + strings.Join(sets.List(idSet), ",")
+			return tags
+		}
+	}
+	return tags
+}
+
+// removeTagOwnership removes a cluster name from the "owned-by-cluster" tag in the tags slice.
+// If the "owned-by-cluster" tag exists, it removes the clusterName from the existing IDs.
+// If the "owned-by-cluster" tag does not exist or the clusterName is not present, it returns the original tags slice unchanged.
+func removeTagOwnership(tags []string, clusterName string) []string {
+	for i, tag := range tags {
+		if strings.HasPrefix(tag, TagPrefixOwnedByCluster) {
+			existingIDs := strings.TrimPrefix(tag, TagPrefixOwnedByCluster)
+			idSet := sets.New[string]()
+			for _, id := range strings.Split(existingIDs, ",") {
+				if id != "" {
+					idSet.Insert(id)
+				}
+			}
+			// Remove the clusterName from the set
+			idSet.Delete(clusterName)
+
+			// If no IDs remain, remove the "owned-by-cluster" tag entirely
+			if idSet.Len() == 0 {
+				return append(tags[:i], tags[i+1:]...)
+			}
+
+			tags[i] = TagPrefixOwnedByCluster + strings.Join(sets.List(idSet), ",")
+			return tags
+		}
+	}
+	return tags
+}
+
+func ownersFromTags(tags []string) []string {
+	for _, tag := range tags {
+		if strings.HasPrefix(tag, TagPrefixOwnedByCluster) {
+			parts := strings.TrimPrefix(tag, TagPrefixOwnedByCluster)
+			ClusterIDs := []string{}
+			for _, id := range strings.Split(parts, ",") {
+				if id != "" {
+					ClusterIDs = append(ClusterIDs, id)
+				}
+			}
+			return ClusterIDs
+		}
+	}
+	return []string{}
+}
+
+// isManagedResource checks if a resource is managed by Kubermatic KKP.
+// Returns true only if:
+// - The resource exists, and It has the tag "managed-by:kubermatic-kkp"
+// Returns false if:
+// - The resource doesn't exist or The tag check fails due to API errors, or The tag is explicitly not present.
+func isManagedResource(netClient *gophercloud.ServiceClient, resourceType, resourceID string) bool {
+	res, _ := tags.Confirm(netClient, resourceType, resourceID, TagManagedByKubermatic).Extract()
+	return res
+}
+
+func addOwnershipToResource(netClient *gophercloud.ServiceClient, resourceType, resourceID, clusterID string) error {
+	resourceTags, err := tags.List(netClient, resourceType, resourceID).Extract()
+	if err != nil {
+		return err
+	}
+	newTags := addTagOwnership(resourceTags, clusterID)
+	return tags.ReplaceAll(netClient, resourceType, resourceID, tags.ReplaceAllOpts{Tags: newTags}).Err
+}
+
+func removeOwnershipFromResource(netClient *gophercloud.ServiceClient, resourceType, resourceID, clusterID string) error {
+	resourceTags, err := tags.List(netClient, resourceType, resourceID).Extract()
+	if err != nil {
+		return err
+	}
+	newTags := removeTagOwnership(resourceTags, clusterID)
+	return tags.ReplaceAll(netClient, resourceType, resourceID, tags.ReplaceAllOpts{Tags: newTags}).Err
```

Comment:
the same here, please don't use RepalceAll, but rather https://pkg.go.dev/github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/attributestags#Delete and only after this List the tags. Otherwise we risk to run into the race condition of 2 competing UCs.